syntax = "proto3";

package org.couchers.api.conversations;

import "google/protobuf/timestamp.proto";
import "google/protobuf/wrappers.proto";
import "google/protobuf/empty.proto";

service Conversations {
  // This is a secure service: a user needs to be authenticated and not jailed to call functions here, refer to auth.proto and jail.proto

  rpc ListGroupChats(ListGroupChatsReq) returns (ListGroupChatsRes) {
    // Retrieves list of group chats (ordered by last message), paginated
  }

  rpc GetGroupChat(GetGroupChatReq) returns (GroupChat) {
    // Retrieves group chat info by id
  }

  rpc GetDirectMessage(GetDirectMessageReq) returns (GroupChat) {
    // Retrieves group chat info for the unique (if any) direct message with a
    // given user
    //
    // Raises: NOT_FOUND if there is no such direct message thread
  }

  rpc GetUpdates(GetUpdatesReq) returns (GetUpdatesRes) {
    // Returns all new messages (aka updates) since a given message id
  }

  rpc GetGroupChatMessages(GetGroupChatMessagesReq) returns (GetGroupChatMessagesRes) {
    // Retrieves messages in group chat, paginated
  }

  rpc MarkLastSeenGroupChat(MarkLastSeenGroupChatReq) returns (google.protobuf.Empty) {
    // Marks the last message that has been seen in this group chat
  }

  rpc CreateGroupChat(CreateGroupChatReq) returns (GroupChat) {
    // Creates a new group chat
  }

  rpc EditGroupChat(EditGroupChatReq) returns (google.protobuf.Empty) {
    // Modifies group chat
  }

  rpc InviteToGroupChat(InviteToGroupChatReq) returns (google.protobuf.Empty) {
    // Invite a person to a chat
  }

  rpc MakeGroupChatAdmin(MakeGroupChatAdminReq) returns (google.protobuf.Empty) {
    // Make a user an admin of a group chat
  }

  rpc RemoveGroupChatAdmin(RemoveGroupChatAdminReq) returns (google.protobuf.Empty) {
    // Remove a user from admin in group chat. TODO: Only original creator can do this
  }

  rpc SendMessage(SendMessageReq) returns (google.protobuf.Empty) {
    // Send a message to a group chat
  }

  rpc LeaveGroupChat(LeaveGroupChatReq) returns (google.protobuf.Empty) {
    // Leave a group chat
  }

  rpc SearchMessages(SearchMessagesReq) returns (SearchMessagesRes) {
    // Search messages by string
  }
}

message MessageContentText {
  string text = 1;
}

message MessageContentChatCreated {}

message MessageContentChatEdited {}

message MessageContentUserInvited {
  uint64 target_user_id = 1;
}

message MessageContentUserLeft {}

message MessageContentUserMadeAdmin {
  uint64 target_user_id = 1;
}

message MessageContentUserRemovedAdmin {
  uint64 target_user_id = 1;
}

enum HostRequestStatus {
  HOST_REQUEST_STATUS_PENDING = 0;
  HOST_REQUEST_STATUS_ACCEPTED = 1;
  HOST_REQUEST_STATUS_REJECTED = 2;
  HOST_REQUEST_STATUS_CONFIRMED = 3;
  HOST_REQUEST_STATUS_CANCELLED = 4;
}

message MessageContentHostRequestStatusChanged {
  //this is the new host request status
  HostRequestStatus status = 1;
}

message Message {
  uint64 message_id = 1;
  uint64 author_user_id = 2;
  google.protobuf.Timestamp time = 3;
  oneof content {
    MessageContentText text = 4;

    MessageContentChatCreated chat_created = 5;
    MessageContentChatEdited chat_edited = 6;
    MessageContentUserInvited user_invited = 7;
    MessageContentUserLeft user_left = 8;
    MessageContentUserMadeAdmin user_made_admin = 9;
    MessageContentUserRemovedAdmin user_removed_admin = 10;
    MessageContentHostRequestStatusChanged host_request_status_changed = 11;
  }
}

message GroupChat {
  uint64 group_chat_id = 1;
  string title = 2;
  repeated uint64 member_user_ids = 3; // inclusive of admins
  repeated uint64 admin_user_ids = 4;
  bool only_admins_invite = 5;
  bool is_dm = 6;
  google.protobuf.Timestamp created = 7;
  uint32 unseen_message_count = 8;
  uint64 last_seen_message_id = 9;
  Message latest_message = 10;
}

message GetGroupChatReq {
  uint64 group_chat_id = 1;
}

message ListGroupChatsReq {
  uint64 last_message_id = 1;
  uint32 number = 2;
}

message ListGroupChatsRes {
  repeated GroupChat group_chats = 1;
  uint64 next_message_id = 2;
  bool no_more = 3;
}

message GetUpdatesReq {
  uint64 newest_message_id = 1;
}

message GetDirectMessageReq {
  uint64 user_id = 1;
}

message Update {
  uint64 group_chat_id = 1;
  Message message = 2;
}

message GetUpdatesRes {
  repeated Update updates = 1;
  bool no_more = 2;
}

message GetGroupChatMessagesReq {
  uint64 group_chat_id = 1;
  uint64 last_message_id = 2;
  uint32 number = 3;
  bool only_unseen = 4;
}

message GetGroupChatMessagesRes {
  repeated Message messages = 1;
  uint64 next_message_id = 2;
  bool no_more = 3;
}

message MarkLastSeenGroupChatReq {
  uint64 group_chat_id = 1;
  uint64 last_seen_message_id = 2;
}

message CreateGroupChatReq {
  google.protobuf.StringValue title = 1;
  repeated uint64 recipient_user_ids = 2;
}

message EditGroupChatReq {
  uint64 group_chat_id = 1;
  google.protobuf.StringValue title = 2;
  google.protobuf.BoolValue only_admins_invite = 3;
}

message InviteToGroupChatReq {
  uint64 group_chat_id = 1;
  uint64 user_id = 2;
}

message MakeGroupChatAdminReq {
  uint64 group_chat_id = 1;
  uint64 user_id = 2;
}

message RemoveGroupChatAdminReq {
  uint64 group_chat_id = 1;
  uint64 user_id = 2;
}

message SendMessageReq {
  uint64 group_chat_id = 1;
  string text = 2;
}

message LeaveGroupChatReq {
  uint64 group_chat_id = 1;
}

message SearchMessagesReq {
  string query = 1;
  uint64 last_message_id = 2;
  uint32 number = 3;
}

message MessageSearchResult {
  uint64 group_chat_id = 1;
  Message message = 2;
}

message SearchMessagesRes {
  repeated MessageSearchResult results = 1;
  uint64 next_message_id = 2;
  bool no_more = 3;
}
